### ISSUE TITLE
Interpretating modified_bases.5mC.bed

### ISSUE THREAD

**Issue by Eric-CH-Chen** [nanoporetech/megalodon#96]
A question about methylation outputs.

I've used Megalodon with the Rerio model (`res_dna_r941_min_modbases_5mC_CpG_v001.cfg`)for a few related fungi species and got some puzzling outputs via the "modified_bases.5mC.bed" output file.
The fungi are from related species (same genus) and cultured and extracted the same way.

| genome assembly size | Num. of 5mC.bed lines | Num > 75% methyl.freq. | Num > 10% methyl.freq. | Num > 1% methyl.freq. |
| --- | --- | ---: | ---: | ---: |
| Sp.1 42 mbp.                         |  25 millions                   | 22     | 15473 | 49132 |
| Sp.2 25 mbp.                         |  15 millions                   | 23354     | 33696 | 95264 |
| Sp.3 40 mbp.                         |  24 millions                   | 36271     | 68361 | 77688 |
| Sp.4 20 mbp.                         |  12 millions                   | 75           | 18078 | 26613  |
| Sp.5 23 mbp.                         |  14 millions                   | 1322     | 19403 | 36702 |
| Sp.6 40 mbp.                         |  24 millions                   | 10390     | 41385 | 83850 |

I understand that the bed doesn't output all the lines in the genome, but why are so many lines simply "0" frequency or extremely low frequency? Is this typical?
Also two of the species (Sp.1, Sp.4, and Sp.5; though especifically sp.1+4) have really low methylation that the algorithm may consider to be real. this seems a little suspicious since the DNA are extracted from live-cultured media and should have a bit more than ~75 positions methylated.

Example command to run megalodon:
```bash
megalodon \
    /data/eric/nanopore_runs/HC_NoStress/Run01_20201130_1108_MN20101_FAO46547_7ef2a872/demutiplexed_fast5/barcode04 \
    --guppy-server-path /home/eric/tools/ont-guppy/bin/guppy_basecall_server \
    --guppy-config res_dna_r941_min_modbases_5mC_CpG_v001.cfg \
    --guppy-params "--disable_pings -d /home/eric/tools/rerio/basecall_models/" \
    --outputs mod_mappings mods \
    --output-directory /home/eric/megalodon_br04_JCM_HC_output/ \
    --haploid \
    --write-mods-text \
    --reference /data/eric/reference_data/JCM_genomes/jcm_assembly_v001.fasta \
    --devices 0 1 2 --processes 2 \
    --overwrite
```

Thanks!

**Comment by marcus1487** [comment]
What is the coverage like for the different samples? Note that the coverage listed in the bedmethyl file will be slightly lower than the raw coverage since reads which do not produce high confidence predictions (modified/canonical) are not included in the reported lines of the bedmethyl output. My guess would be that the lower coverage may be causing some of these issues.

I will also note that the 5mC model performs very well on CpG contexts, but does not perform quite as well in other contexts. We are working to improve this feature, but this may also be contributing to these results.

**Comment by Eric-CH-Chen** [comment]
Thanks for the speedy response! Ah ok so the lines reporting 0% methylation frequency likely had some reads that did not have high confidence and were filtered out? Ok that makes sense.

The coverage ranges from 13x (Sp. 3) to a little over 50x (Sp2) and the rest are in the 15-28X range.  For Species 1. I tested a run by combining 3 biological replicates, pushing the coverage to 76X and still the similar output (table below).

Currently rerunning the unmerged dataset on CpG context ` --mod-motif m CG 0 ` option added to see if there would be significant difference.

| Species | Num. Bed lines | Num. > 75% | Num. > 10% | Num. > 1% |
| --- | ---: | ---: | ---: | ---: | 
| Sp.3 | 25 millions | 3 | 1830 | 163664|

**Comment by marcus1487** [comment]
For the `--mod-motif` option there will be no difference to calls made at CpG sites. This flag simply filters the amount of data produced and processing required to make megalodon faster and use less disk space.

I would recommend instead that you can use the `megalodon_extras modified_bases create_motif_bed` command to filter results by motif with an associated `bedtools intersect` command.

For the coverage results, are these raw coverage or coverage obtained from the bedmethyl files? The coverage listed in the bedmethyl files is the "valid" coverage (with confident modified/canonical calls) contributing to the fraction modified output.

**Comment by Eric-CH-Chen** [comment]
Thanks for the reply.  the `--mod-motif` explanation makes... too much sense haha.

For coverage, it is based on guppy's HAC model basecalls that passed the Q=7 threshold. My samples were multiplexed so I used the guppy's basecall result to subset the raw fast5s into individual barcodes before feeding them to megalodon. 

Also, a side question, for the coverage in bedmethyl file, would I expect the it to cover the entire length of the assembly if enough coverage was sequenced with high enough quality?

For the `megalodon_extras modified_bases create_motif_bed` command suggestion, wouldn't I still not find much? Since the output bedMethyl file from the megalodon wouldn't consider those positions to be methylated (more canonical calls than modified calls)?

Thank you!

**Comment by marcus1487** [comment]
Megalodon calls will cover all sites covered by mapped reads. In most cases calls with lower probabilities (thus filtered out) are not biased to positions in the reference or within particular reads. Therefore I would expect the bedmethyl calls to include all sites with sufficient coverage.

For the motif bed this was a suggestion in order to obtain the calls at CpG sites without re-running megalodon (and the computationally expensive basecalling step). The calls output when setting the `--mod-motif` option are the exact same as those without setting this option (this option just filters these sites to avoid some unnecessary processing and storage).

**Comment by Eric-CH-Chen** [comment]
Sorry for the late response and thank you again for the explanation.

So as far as my data/results goes, I should expect the following to be true?

1. I am not messing up parameters for megalodon and it is calling methylation as designed. 
2. Broad support (via methylation freq) positions to be methylated across the sequenced sample (at least in Species 1, 4 and 5) are not there because it really is not there. (ie. Sp.1 really only have 22 positions to be relatively confident 5mC)

And/or

3. Samples that I sequenced may be very heterozygous so I am unable to find broad support for genomic positions to be methylated.

Final question (hopefully!), is it recommended to quality filter the reads before feeding the raw_fast5 to megalodon? ie. Call bases first via guppy, then subset the raw fast5 to remove any reads less than X length and Q of 7. Before finally input it to megalodon for methylation detection. If it is, is it possible to pass quality filters via megalodon directly?

Thank you so much for your help!

**Comment by marcus1487** [comment]
1. The parameters listed in the command and the rerio model used are as recommended. I don't see any issues that would cause issues with methylation calling.
2. There may still be some issues with 5mC calling outside of CpG contexts that could lead to false negative methylation calls. Could you have a look at the coverage values reported in the bedmethyl file? These represent the coverage contributing to the fraction modified estimates and may be much lower due to uncertain calls across the genome.
3. I'm not sure exactly what the question is for this point. Could you rephrase this assumption?

For the last question, we generally do not recommend filtering the raw fast5. There is not a q-score or length filter in the standard megalodon command. If such filters were required, I would recommend running megalodon with the settings `--outputs mappings mods`. The mappings summary output file will contain the q-score and the mapped length. This file could be filtered as required. This filtered set of read ids could then be passed to the `megalodon_extras aggregate run` command to only produce aggregated modified base calls from reads passing filters.

**Comment by Eric-CH-Chen** [comment]
Thanks for the speedy response!

On point 3. Sorry for the confusion. What I meant is that perhaps my sample is heterogenous in the sense that the different individuals are all doing different things; resulting in very different methlations patterns. So when I average across the methylation profile across all individuals, almost none of the positions are supported by the majority of reads to be methylated.  This is more or a potential possibiliy to explain the general lack of methylation in some of my samples.

Based on the bedmethyl file
| Species | Avg. Coverage  (Bedmethyl - all) | Expected Coverage (Guppy Q7.fil.) |
| --- | ---: | ---: |
| Sp.1 | 11 | 28 |
| Sp.2 | 23 | 53 |
| Sp.3 | 7 | 18 |
| Sp.4 | 7 | 15 |
| Sp.5 | 10 | 22 |
| Sp.6 | 11 | 25 |

Since there is a pretty consistent "drop" of ~50% of the coverage being useful, I suppose this is working as intended?

**Comment by marcus1487** [comment]
The explanation for point 3 certainly would make sense. I don't know enough about fungal biology to make any comments on that.

For the coverage, methylation occurs on a single strand (forward or reverse strand). Thus the drop in coverage is due to the calls being made only on the reads mapping to the strand being called (not a lack of confident calls). This does indicate that megalodon is making confident calls at most all sites. This does not rule out that the calls in non-CG contexts are not false negatives and there is still some methylation here. We are working hard to expand our training datasets and improve the methylation detection performance outside of CG contexts. Keep an eye out for updates on this front (models will be added to the Rerio repository initially when released).
